### DISCUSSION TITLE
How to receive OS Push Notification in Android app?

### DISCUSSION THREAD

**Discussion by williamspete321** [Iterable/iterable-android-sdk#484] [Q&A] [answered]
If we select to use OS Notification, how can we set up our Android app to receive this notification from the Android OS? I assume we want to catch it in a [Broadcast Receiver](https://developer.android.com/guide/components/broadcasts). Is there an action name that we define in a receiver in the Android Manifest to catch the Iterable notification?

`<receiver android:name=".MyBroadcastReceiver" android:exported="false">
    <intent-filter>
        <action android:name="ITERABLE_SPECIFIC_BROADCAST_HERE" />
    </intent-filter>
</receiver>`

**Comment by Ayyanchira** [comment, marked as answer]
Hi @williamspete321! Thanks for posting the question. Pre-defined action names are used with notifications created by IterableSDK to be used by Iterable's BroadcastReceiver. Can you share the use case you are trying to solve with OS notification?

**Reply by williamspete321** [reply]
Hi @Ayyanchira. Yes, I'd like to send raw data through JSON in the Iterable push campaign. If I let the Iterable SDK handle the push notification, I can't see any action name or data in the Intent object that our app receives from Iterable. Therefore, I want to disable the Iterable SDK and let the Android OS handle the notification that Iterable sends to the device. In order for our app to catch this Intent, we need to know things like the action name to listen for. Could you let me know what the pre-defined action names are that are used with notifications sent by Iterable so we can create a Broadcast Receiver and Intent Filter to catch these Intent objects? I couldn't find these specifics in the Iterable Android documentation.

**Reply by williamspete321** [reply]
I tried adding the following in our app's manifest, but so far I'm not catching any Iterable Push notification:

`<receiver android:name=".MyBroadcastReceiver" android:exported="false">
<intent-filter>
     <action android:name="com.iterable.push.ACTION_PUSH_ACTION" />
</intent-filter>
</receiver>`

**Reply by Ayyanchira** [reply]
Hi @williamspete321, We recommend using SDK for rendering notification as it has many benefits like rich push and tracking. What functionality do you intend achieve using raw json data?

**Reply by williamspete321** [reply]
Hi @Ayyanchira, ideally we want our push notifications to be as customizable as possible. Sending raw json data gives us this ability. Does my previous question about Broadcast Receiver's make sense?

**Reply by Ayyanchira** [reply]
@williamspete321 , yes the question does make sense, but Im trying to understand further. OS notifications provide no customization. Do you intend to create notification using raw data?

**Reply by williamspete321** [reply]
@Ayyanchira We will handle the notification creation and customization ourself inside the Receiver class. But to first do that we need to catch the Intent object. If I send a Push Notification then the example Intent Filter I posted earlier does not catch this Intent. That is why I'm asking what the `action` name of the Iterable Push notification and if the `intent-filter` above is correct? Because we can't catch the Intent, I'm wondering if something is missing. Only when using the Iterable SDK do we catch the Intent in our main app. Because the Iterable web console gives us the option to let the Android OS handle the notification, I think there needs to be some documentation regarding what the name of the notification is that our Android OS needs to look out for.

**Reply by williamspete321** [reply]
"Do you intend to create notification using raw data?" Yes, we intend to create the notification ourselves with the Intent object we catch in the Receiver and any data that it contains. So we just need to dig a little and understand how to catch this Intent when we choose not to handle the Intent (notification) with the Iterable SDK.

**Reply by Ayyanchira** [reply]
For this use case, suggestion will be to NOT use OS notification. 
To create custom notification using raw json data, we would recommend creating one with your own custom `FirebaseMessagingService` implementation. 
This will allow app to intercept firebase messages sent from Iterable along with raw data. Raw data is accessible via `remoteMessage.getData()`
Intents to be created for any service or activity can be stored at this point before notification is posted. 
Hope this helps.

**Reply by williamspete321** [reply]
Thanks @Ayyanchira, this is working nicely and we've been able to intercept the RemoteMessage object and build our own notification.
